backend/lib/kubernetes-client/resources/Resource.js

```diff
@@ -0,0 +1,167 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+'use strict'
+
+const { URLSearchParams } = require('url')
+const { join } = require('path')
+const { EventEmitter } = require('events')
+const pRetry = require('p-retry')
+const pEvent = require('p-event')
+const WebSocket = require('ws')
+const inject = require('reconnect-core')
+
+const BaseResource = require('./BaseResource')
+const { http, ws } = require('./symbols')
+
+const { setPatchType, wrapWebSocket, waitFor } = require('../util')
+
+class Resource extends BaseResource {
+  constructor (options = {}) {
+    super({ ...options, responseType: 'json' })
+  }
+
+  [http.prefixUrl] (url) {
+    let prefixUrl = super[http.prefixUrl](url)
+    const { group, version } = this.constructor
+    if (group === 'core' || group === '') {
+      prefixUrl += '/api/' + version
+    } else {
+      prefixUrl += '/apis/' + group + '/' + version
+    }
+    return prefixUrl
+  }
+
+  [http.pathname] ({ namespace, allNamespaces, name }) {
+    let pathname = ''
+    if (this.constructor.scope === 'Namespaced') {
+      if (allNamespaces) {
+        if (name) {
+          throw new TypeError('A namespaced resource cannot be addressed by name across all namespaces')
+        }
+      } else if (!namespace) {
+        throw new TypeError('Either \'namespace\' or \'allNamespaces\' must be specified for a namespaced resource')
+      } else {
+        pathname += 'namespaces/' + encodeURIComponent(namespace) + '/'
+      }
+    }
+    pathname += this.constructor.names.plural
+    if (name) {
+      pathname += '/'
+      pathname += Array.isArray(name) ? name.map(encodeURIComponent).join('/') : encodeURIComponent(name)
+    }
+    return pathname
+  }
+
+  [http.patch] ({ type = 'merge', ...options } = {}) {
+    return super[http.patch](setPatchType(options, type))
+  }
+
+  [ws.create] ({ allNamespaces, namespace, name, query, headers, origin, ...options } = {}) {
+    const {
+      prefixUrl,
+      ca,
+      key,
+      cert,
+      servername,
+      rejectUnauthorized,
+      headers: defaultHeaders
+    } = this[http.client].defaults.options
+    const url = new URL(prefixUrl)
+    origin = origin || url.origin
+    url.protocol = url.protocol.replace(/^http/, 'ws')
+    url.pathname = join(url.pathname, this[http.pathname]({ allNamespaces, namespace }))
+    const searchParams = new URLSearchParams(query)
+    searchParams.set('watch', true)
+    if (name) {
```

Comment:
introduce allNames param, similar to allNamespaces

Reply:
✔️